src/server/worker/datum_iterator.go

```diff
@@ -0,0 +1,295 @@
+// Note that since you start the loop by a call to Next(), the datum iterator's location starts at -1
+type DatumIterator interface {
+	Reset()
+	Len() int
+	Next() bool
+	Datum() []*Input
+}
+
+type pfsDatumIterator struct {
+	inputs   []*Input
+	location int
+}
+
+func newPFSDatumIterator(pachClient *client.APIClient, input *pps.PFSInput) (DatumIterator, error) {
+	result := &pfsDatumIterator{}
+	if input.Commit == "" {
+		// this can happen if a pipeline with multiple inputs has been triggered
+		// before all commits have inputs
+		return result, nil
+	}
+	fs, err := pachClient.GlobFileStream(pachClient.Ctx(), &pfs.GlobFileRequest{
+		Commit:  client.NewCommit(input.Repo, input.Commit),
+		Pattern: input.Glob,
+	})
+	if err != nil {
+		return nil, err
+	}
+	for {
+		fileInfo, err := fs.Recv()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, &Input{
+			FileInfo:   fileInfo,
+			Name:       input.Name,
+			Lazy:       input.Lazy,
+			Branch:     input.Branch,
+			EmptyFiles: input.EmptyFiles,
+		})
+	}
+	// We sort the inputs so that the order is deterministic. Note that it's
+	// not possible for 2 inputs to have the same path so this is guaranteed to
+	// produce a deterministic order.
+	sort.Slice(result.inputs, func(i, j int) bool {
+		// We sort by descending size first because it can boost performance to
+		// process the biggest datums first.
+		if result.inputs[i].FileInfo.SizeBytes != result.inputs[j].FileInfo.SizeBytes {
+			return result.inputs[i].FileInfo.SizeBytes > result.inputs[j].FileInfo.SizeBytes
+		}
+		return result.inputs[i].FileInfo.File.Path < result.inputs[j].FileInfo.File.Path
+	})
+	// make sure it gets initialized properly
+	result.Reset()
+	return result, nil
+}
+
+func (d *pfsDatumIterator) Reset() {
+	d.location = -1
+}
+
+func (d *pfsDatumIterator) Len() int {
+	return len(d.inputs)
+}
+
+func (d *pfsDatumIterator) Datum() []*Input {
+	return []*Input{d.inputs[d.location]}
+}
+
+func (d *pfsDatumIterator) Next() bool {
+	d.location++
+	return d.location < len(d.inputs)
+}
+
+type unionDatumIterator struct {
+	inputs   []DatumIterator
+	unionIdx int
+}
+
+func newUnionDatumIterator(pachClient *client.APIClient, union []*pps.Input) (DatumIterator, error) {
+	result := &unionDatumIterator{}
+	for _, input := range union {
+		datumIterator, err := NewDatumIterator(pachClient, input)
+		if err != nil {
+			return nil, err
+		}
+		result.inputs = append(result.inputs, datumIterator)
+	}
+	result.Reset()
+	return result, nil
+}
+
+func (d *unionDatumIterator) Reset() {
+	for _, input := range d.inputs {
+		input.Reset()
+	}
+	d.unionIdx = 0
+}
+
+func (d *unionDatumIterator) Len() int {
+	result := 0
+	for _, datumIterator := range d.inputs {
+		result += datumIterator.Len()
+	}
+	return result
+}
+
+func (d *unionDatumIterator) Next() bool {
+	if len(d.inputs) == 0 {
+		return false
+	}
+	inputHasNext := d.inputs[d.unionIdx].Next()
```

Comment:
Is this correct? I don't think this would work as intended. It looks like it returns false when the first inner Next returns true, and does not seem like it would handle the case when the second inner Next returns false (0 inputs) but is followed by additional datum iterators. If I am understanding this correctly, I think it might make sense to refactor this to be recursive. So, you could have something like this: 
```
if d.unionIdx >= len(d.inputs) {
  return false
}
if !d.inputs[d.unionIdx].Next() {
  d.unionIdx++
  return d.Next()
}
return true
```
Also, it might make sense to rename `d.inputs` to `d.iterators` or something to indicate they are iterators since it is a little confusing to overload the word inputs.